Content.Shared/_RMC14/Attachable/AttachableHolderSystem.cs

```diff
@@ -279,38 +281,48 @@ private void OnAttachDoAfter(EntityUid uid, AttachableHolderComponent component,
         var container = _container.EnsureContainer<ContainerSlot>(holder, slotId);
         container.OccludesLight = false;
 
-        if (container.Count > 0 && !Detach(holder, attachableUid, userUid, slotId))
+        if (container.Count > 0 && !Detach(holder, container.ContainedEntities[0], userUid, slotId))
             return false;
 
         if (!_container.Insert(attachableUid, container))
             return false;
-
-        UpdateStripUi(holder.Owner, holder.Comp);
-
-        var holderEv = new AttachableHolderAttachablesAlteredEvent(attachableUid, slotId, AttachableAlteredType.Attached);
-        RaiseLocalEvent(holder, ref holderEv);
-
-        var ev = new AttachableAlteredEvent(holder.Owner, AttachableAlteredType.Attached);
-        RaiseLocalEvent(attachableUid, ref ev);
-
-        var addEv = new GrantAttachableActionsEvent(userUid);
-        RaiseLocalEvent(attachableUid, ref addEv);
-
-        _gun.RefreshModifiers(holder.Owner);
-
+        
+        if(_hands.IsHolding(userUid, holder.Owner))
+        {
+            var addEv = new GrantAttachableActionsEvent(userUid);
+            RaiseLocalEvent(attachableUid, ref addEv);
+        }
+        
         _audio.PlayPredicted(Comp<AttachableComponent>(attachableUid).AttachSound,
             holder,
             userUid);
 
-        Dirty(holder);
-
         return true;
     }
+    
+    private void OnAttached(Entity<AttachableHolderComponent> holder, ref EntInsertedIntoContainerMessage args)
+    {
+        if (!HasComp<AttachableComponent>(args.Entity) || !holder.Comp.Slots.ContainsKey(args.Container.ID))
+            return;
+        
+        if (args.OldParent == holder.Owner)
+            return;
+        
+        UpdateStripUi(holder.Owner, holder.Comp);
+
+        var ev = new AttachableAlteredEvent(holder.Owner, AttachableAlteredType.Attached);
+        RaiseLocalEvent(args.Entity, ref ev);
+        
+        var holderEv = new AttachableHolderAttachablesAlteredEvent(args.Entity, args.Container.ID, AttachableAlteredType.Attached);
+        RaiseLocalEvent(holder, ref holderEv);
+        
+        Dirty(holder);
+    }
 
     //Detaching
     public void StartDetach(Entity<AttachableHolderComponent> holder, string slotId, EntityUid userUid)
     {
-        if (TryGetAttachable(holder, slotId, out var attachable))
+        if (TryGetAttachable(holder, slotId, out var attachable) && !holder.Comp.Slots[slotId].Locked)
```

Comment:
Is Slots guaranteed to have a slot with slotId here? Feels like it should be a try get instead specially since this is a public method.

Reply:
It is *not*. Good catch, I completely missed this somehow. Fixed now.